VSD pointer arithmetic, differencing, and comparisons

jezhiggins opened the pull request:
Pointer arithmetic fixes 

- Evaluates `(p - q)` correctly, when `p` and `q` point to the same base object
- Pointer manipulation evaluates correctly when the result points to the base object. eg 
``` 
int* p = array;
++p;
--p;
assert (p == array);
```
Assertion is now SUCCESS, when previously it was UNKNOWN.

----
A pointer difference or comparison expression is slightly unusual - the type of the expression is a number (or boolean), but the two operands are pointer types. Generally, the expression type is the same as the first operand (if not all of them). Because of this type mismatch, the abstract environment was dispatching the expression to the wrong abstract type for evaluation, so the result was TOP and therefore unhelpful.

The changes to evaluate pointer differences are small - an additional case in `abstract_environmentt::eval` to pick up a pointer difference and dispatch it to `abstract_pointer_object`, and then the corresponding changes in `abstract_pointer_object` to calculate the difference. 

Pointer comparisons are a little more involved. Again, we have an additional case in `abstract_environmentt::eval` to dispatch the expression to `abstract_pointer_object`. If the two pointers point to the same base object we can rewrite the comparison in terms of the pointer offset and evaluate that. If the two pointers point to different objects, we can evaluate == and !=, but other comparisons go top.

jezhiggins committed:
Convert pointer difference expressions from minus_exprt to ptr_diff_exprt

jezhiggins committed:
Dispatch ptr_diff_exprts to lhs object for evaluation

jezhiggins committed:
ptr_diff_t expression evaluation

jezhiggins committed:
ptr_diff_exprt evaluation for constant_pointer_abstract_object

jezhiggins committed:
Pointer differencing and arithmetic tests

jezhiggins committed:
Don't simplify (T*)p+a when a is a number.

If a pointer addition simplifies to p+a, if we continue to simplify we
risk incorrectly eliminating zeros from the expression, erroneously
going to TOP when we would otherwise have a correct pointer to the base
object.

jezhiggins committed:
Add switch to make pointer simplication changes VSD only

jezhiggins commented:
**Build Xen with CPROVER tools / CompileXen (pull_request)** is failing for unrelated reasons.

```
Fetched 6178 kB in 1s (5261 kB/s)
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/glibc/libc6-dbg_2.27-3ubuntu1.3_amd64.deb  404  Not Found [IP: 52.147.219.192 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

martin-cs commented on regression/goto-analyzer/pointer-arithmetic-simple/test.desc:
```diff
@@ -0,0 +1,7 @@
+CORE
+main.c
+--variable-sensitivity --vsd-pointers constants --vsd-values constants --verify --pointer-check
```
Why is `--pointer-check` needed?

martin-cs commented on regression/goto-analyzer/pointer-comparison/main.c:
```diff
@@ -0,0 +1,15 @@
+  assert(p - q == 0);
+
+  q = b;
+  assert(p - q == 0);
```
I have a feeling that the C standard says this is undefined behaviour.  We probably should support it and so it is reasonable to test for it but until we have a more principled way of handling UB, we should be careful.  If you have to modify the code then a comment noting that this is intentionally UB would be good.

martin-cs commented on regression/goto-analyzer/pointer-difference-after-decrement/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=0$
+^SIGNAL=0$
+^\[main.assertion.1\] line 10 assertion q - p == 1: SUCCESS
+^\[main.assertion.2\] line 11 assertion q != p: UNKNOWN
```
Why is this unknown?  If the previous one is resolvable it feels like it should be.

jezhiggins replied:
Turns out the `q-p == 1` case evaluating to success was an artefact. I'm implementing proper pointer comparisons and `q != p` will evaluate properly.

martin-cs commented on regression/goto-analyzer/pointer-difference-after-decrements/main.c:
```diff
@@ -0,0 +1,29 @@
+  offset = (q - p);
+  assert(offset == 0);
+  assert(q == p);
+}
```
Nice test!

martin-cs commented on regression/goto-analyzer/pointer-comparison/test.desc:
```diff
@@ -0,0 +1,7 @@
+CORE
+main.c
+--variable-sensitivity --vsd-pointers constants --vsd-values constants --verify --pointer-check
```
Why `--pointer-check`?

martin-cs commented on regression/goto-analyzer/pointer-difference-after-increment/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=0$
+^SIGNAL=0$
+^\[main.assertion.1\] line 10 assertion q - p == 1: SUCCESS
+^\[main.assertion.2\] line 11 assertion q != p: UNKNOWN
```
Again, somewhat unclear why this is unknown.

martin-cs commented on regression/goto-analyzer/pointer-differences/test.desc:
```diff
@@ -0,0 +1,12 @@
+^\[main.assertion.3\] line 20 assertion offset == 1: SUCCESS$
+^\[main.assertion.4\] line 22 assertion offset == 0: SUCCESS$
+^\[main.assertion.5\] line 26 assertion offset == 2: SUCCESS$
+^\[main.assertion.6\] line 27 assertion q == p: UNKNOWN$
```
Another unclear one; equality should be resolvable to a difference check if nothing else.

martin-cs commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -517,3 +519,18 @@ std::vector<abstract_object_pointert> eval_operands(
   return operands;
 }
+
+class simplify_vsd_exprt : public simplify_exprt
```
Having a custom simplification for VSD is not unreasonable but I am not immediately seeing what it adds in this case.

jezhiggins replied:
It's gone!

martin-cs commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -70,6 +71,35 @@ constant_pointer_abstract_objectt::merge(abstract_object_pointert other) const
+    std::dynamic_pointer_cast<const constant_pointer_abstract_objectt>(other);
+
+  if(value_stack.is_top_value() || cast_other->value_stack.is_top_value())
+    return nil_exprt();
```
I would really like to get rid of the use of `nil_exprt()` as it feels very kludged.  Can this method return an `abstract_objectt` so it has a way of returning top.  Failing that?  `optionalt`?

jezhiggins replied:
This is calling site is `return environment.eval(offset_from(rhs), ns);` so the `nil_exprt()` does end up as a TOP value.

martin-cs commented on src/analyses/variable-sensitivity/pointer_abstract_object.h:
```diff
@@ -97,6 +97,15 @@ class pointer_abstract_objectt : public abstract_objectt
 protected:
   CLONE
 
+  virtual bool same_target(abstract_object_pointert other) const
```
I am not entirely opposed to adding to the interface for pointers if it is necessary.  However there is a general move towards making the interface of `abstract_objectt` and children more uniform so I wonder if there is another way.

martin-cs commented on src/util/irep_ids.def:
```diff
@@ -842,6 +842,7 @@ IREP_ID_ONE(statement_list_instructions)
 IREP_ID_ONE(max)
 IREP_ID_ONE(min)
 IREP_ID_ONE(constant_interval)
+IREP_ID_ONE(ptr_diff)
```
I realise it is quite an elegant solution to the problem but I would ideally like to avoid adding an `irept` type that is a restricted version of another type / echos some of the type information.  I haven't got that far through the PR, maybe there is something that is simply not possible without it but, could we just have a utility method for `is this subtract a pointer difference`?

jezhiggins replied:
Yep - this and all the scaffolding to support it is gone.

martin-cs reviewed (changes requested):
I like what this is doing, there are a few places where I feel how it is doing it could be better.  I suspect "better" here means deeper and more intrusive changes to the `write_stack` parts and `constant_pointer_abstract_objectt` which are only obviously "better" when taking into account a load of other things so why don't we have a chat and see what can be done?

jezhiggins committed:
Remove ptr_diff_exprt

Detect case by checking expr has ID_minus and the operands are both
pointers

jezhiggins committed:
Pulled functions out of abstract_pointer_object::expression_transform

jezhiggins committed:
Additional pointer comparison regression test

jezhiggins committed:
Clarifying helper functions in abstract_environment::eval

jezhiggins committed:
Dispatch pointer comparisons to abstract_pointer_object

jezhiggins committed:
Evaluate pointer comparisons, when both pointers have same target

jezhiggins committed:
Move behaviour out of abstract_pointer_object

It implements some of what should be two_value_pointer_abstract_object,
so moved that out.

jezhiggins committed:
Make abstract_pointer_object::ptr_comparison_expr pure virtual.

Representation specific implementations in derived classes.
value_set_pointer currently a stub.

jezhiggins committed:
pointer comparisons for value-set of pointers

jezhiggins committed:
Fill out pointer comparison for different targets

jezhiggins committed:
Pointer to struct member comparisons

jezhiggins committed:
comparisons of pointers to symbols

jezhiggins committed:
value-set pointer differences

jezhiggins committed:
same_target, offset, offset_from can be private to constant abstract pointer

jezhiggins committed:
Remove duplication in ptr_comparison_expr

martin-cs commented on regression/goto-analyzer/pointer-comparison-equality-struct-members/main.c:
```diff
@@ -0,0 +1,28 @@
+    int a;
+    int b;
+  };
+  struct int_pair x = {0, 1.0};
```
Is the use of a double here significant?

jezhiggins replied:
No. I originally wrote `struct { int a, float b; };` but then changed it. Can#t remember why.

martin-cs commented on regression/goto-analyzer/pointer-comparison-inequality-different-target/main.c:
```diff
@@ -0,0 +1,20 @@
+
+  int *p = &(a[0]);
+  int *q = &(b[1]);
+
```
Can we have a comment in either in the test case or in the `.desc` files noting that this is undefined behaviour and so `UNKNOWN` is the expected behaviour.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -824,3 +854,18 @@ exprt assume_greater_than(
   return assume_less_than(env, symmetric_expr, ns);
 }
+
+class simplify_vsd_exprt : public simplify_exprt
```
I would rather not do this, if possible.

jezhiggins replied:
We do have to do it, but I've made further changes to try and make things less intrusive.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -53,59 +54,88 @@ std::vector<abstract_object_pointert> eval_operands(
+bool is_ptr_comparison(const exprt &expr)
+{
+  auto const &id = expr.id();
+  if(id == ID_not)
```
Are you handling `ID_and` elsewhere?  If so, should you handle `ID_not` there rather than here?

jezhiggins replied:
Yep, sorted

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -53,59 +54,88 @@ std::vector<abstract_object_pointert> eval_operands(
 
   // first try to canonicalise, including constant folding
-  const exprt &simplified_expr = simplify_expr(expr, ns);
+  const exprt &simplified_expr = simplify_vsd_expr(expr, ns);
```
Pondering : do we have to simplify here?  Could we do special case recognition first?

jezhiggins replied:
We have to customise the simplification, because it recurses right down the expression tree. If we just let the simplification run, then the special case ends up being eliminated before we get a chance to try and evaluate it.

martin-cs commented on src/analyses/variable-sensitivity/abstract_pointer_object.cpp:
```diff
@@ -74,48 +73,79 @@ abstract_object_pointert abstract_pointer_objectt::write(
 }
 
-abstract_object_pointert abstract_pointer_objectt::read_dereference(
+void abstract_pointer_objectt::get_statistics(
```
FYI : I am not sure whether this infrastructure is used at the moment.

martin-cs commented on src/analyses/variable-sensitivity/abstract_pointer_object.cpp:
```diff
@@ -74,48 +73,79 @@ abstract_object_pointert abstract_pointer_objectt::write(
 {
-  abstract_objectt::get_statistics(statistics, visited, env, ns);
-  ++statistics.number_of_pointers;
+  if(expr.id() == ID_not)
```
It feels rather like breaking conditions into atoms should be done elsewhere.

jezhiggins replied:
Maybe I got carried away here. I can unwind this.

martin-cs commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -72,6 +72,48 @@ abstract_object_pointert constant_pointer_abstract_objectt::merge(
+  if(value_stack.depth() != cast_other->value_stack.depth())
+    return false;
+
+  for(size_t d = 0; d != value_stack.depth() - 1; ++d)
```
Looks reasonable, the -1 is needed because you don't want the most detailed part of the offset.

martin-cs commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -72,6 +72,48 @@ abstract_object_pointert constant_pointer_abstract_objectt::merge(
+
+  return minus_exprt(
+    value_stack.offset_expression(),
+    cast_other->value_stack.offset_expression());
```
Why convert back to exprt's when you could just subtract the two abstract objects?

jezhiggins replied:
I'm not :)

martin-cs commented on src/analyses/variable-sensitivity/context_abstract_object.cpp:
```diff
@@ -86,8 +86,6 @@ abstract_object_pointert context_abstract_objectt::expression_transform(
   const abstract_environmentt &environment,
   const namespacet &ns) const
 {
-  PRECONDITION(expr.operands().size() == operands.size());
```
Why?

martin-cs commented on src/util/std_expr.h:
```diff
@@ -922,22 +914,31 @@ inline const minus_exprt &to_minus_expr(const exprt &expr)
   {
   }
+
+  const exprt &lhs() const
```
This is potentially a bit misleading as `mult_exprt` can have more than 2 operands.

martin-cs reviewed (changes requested):
The actual content of the PR looks good.

Reformatting of things in `util/` will need to go in a separate PR if they are needed.

jezhiggins committed:
let inverted pointer expression eval as normal. No need for special case.

jezhiggins committed:
Unwind vsd customisation from simply_exprt.

We can pick up the special case for pointer addition in
abstract_object::expression_transform.

martin-cs reviewed (approved):
Thanks for removing the changes to `util/` I think it is a better PR for the extra passes.
